### PR TITLE
WIP: Create library for parsing commandline flags -> protocol buffer, decouple cobra

### PR DESCRIPTION
**Pull request by shiywang**
Fixes https://github.com/kubernetes/kubectl/issues/94
 
/assign @pwittrock 
still need some tests, and I'm not sure if this is what we want, @pwittrock  please leave some comments, also currently the protofile generates by command `protoc --go_out=. ./cmd.proto` manually.

I also have some other ideas about decoupled cobra, will file an issue later. 

I rewrite the command `version` flag setup in new way(protobuf), you could check the `version.go` file to see if this is decoupled enough.

also @kubernetes/sig-cli-feature-requests @kubernetes/sig-cli-maintainers @kubernetes/sig-cli-misc 

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/cmd/util/cmdproto/cmdproto_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+			name:        "test with abnormal value",
+			flagName:    []string{"bool-flag", "int32-flag", "int64-flag", "time-flag"},
+			flagValue:   []string{"not-true", "not-int32", "not-int64", "not-time"},
+			expectedErr: []bool{true, true, true, true},
+			expectedOut: Test{
+				BoolFlag:   false,
+				StringFlag: "",
+				Int32Flag:  0,
+				ArrayFlag:  []string{"default1", "default2"},
+				Int64Flag:  0,
```
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdproto/cmdproto_test.go]: expect to see 1234, but this maybe a bug for cobra, file an issue here: https://github.com/spf13/cobra/issues/557

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+message Array {
+    repeated string array = 1;
+}
+
+//Flags Description
+
+
+//This is the flag description for `kubectl version`
+message VersionFlags {
+  optional bool client = 2 [ (info) = { name: "client" value: "false" shorthand: "c" usage: "Client version only (no server required)."}];
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: You can specify the default value in a type-safe way.
https://developers.google.com/protocol-buffers/docs/proto#optional

File: pkg/kubectl/cmd/util/cmdproto/cmdproto.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
+package cmdproto
+
+import (
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/cmdproto.go]: how about moving this to: k8s.io/kubectl/pkd/framework/proto?

File: pkg/kubectl/cmd/util/cmdproto/cmdproto.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		return "", "", []string{""}, "", cmdproto.FlagDetail_NONE, err
+	}
+	return info.(*cmdproto.FlagDetail).GetName(), info.(*cmdproto.FlagDetail).GetShorthand(), info.(*cmdproto.FlagDetail).GetValue(), info.(*cmdproto.FlagDetail).GetUsage(), info.(*cmdproto.FlagDetail).GetExt(), nil
+}
+
+func FlagsSetup(cmd *cobra.Command, msg interface{}) {
+	v := reflect.ValueOf(msg).Elem()
+
+	for i := 0; i < v.NumField()-1; i++ {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/cmdproto.go]: What would this look like with proto reflection instead of go reflection?

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+message Array {
+    repeated string array = 1;
+}
+
+//Flags Description
+
+
+//This is the flag description for `kubectl version`
+message VersionFlags {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: Lets just create a test example for this library without trying to replace existing commands for it.

Rename this `TestCmd`.

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  enum EXT {
+      NONE = 0;
+      ISTIME = 1;
+      ISSTRARRAY = 2;
+      ISSTRSLICE = 3;
+  }
+  optional EXT ext = 5;
+}
+
+extend google.protobuf.FieldOptions {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: Lets add another for message options

```proto
extend google.protobuf.MessageOptions {
  optional string description = 51234;
}

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+//This is the flag description for `kubectl version`
+message VersionFlags {
+  optional bool client = 2 [ (info) = { name: "client" value: "false" shorthand: "c" usage: "Client version only (no server required)."}];
+  optional bool short = 3  [ (info) = { name: "short" value: "false" shorthand: "" usage: "Print just the version number."}];
+  optional string output = 4 [ (info) = { name: "output" value: "" shorthand: "o" usage: "One of 'yaml' or 'json'."}];
+}
+
+//This is the flag description only for test
+message TestTest {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: Lets call it `TestCmd` instead

Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: WDYT of adding support for parsing the command description and example as options as well.

File: pkg/kubectl/cmd/version.go [hunk 8]
```diff
@@ -28,10 +28,18 @@ import (
 	apimachineryversion "k8s.io/apimachinery/pkg/version"
 	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
 	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/util/cmdproto"
+	protofile "k8s.io/kubernetes/pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd"
 	"k8s.io/kubernetes/pkg/kubectl/util/i18n"
 	"k8s.io/kubernetes/pkg/version"
 )
 
+var (
```
Comment (pwittrock) [inline pkg/kubectl/cmd/version.go]: Lets not convert and existing command yet.  I would like to see how this plays out in a green field environment first.

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+import "google/protobuf/descriptor.proto";
+
+message FlagDetail {
+  required string name = 1;
+  repeated string value = 2;
+  required string shorthand = 3;
+  required string usage = 4;
+  enum EXT {
+      NONE = 0;
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: Why are these necessary?

File: pkg/kubectl/cmd/util/cmdproto/cmdproto.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	_, md := descriptor.ForMessage(msg.(descriptor.Message))
+	opts := md.Field[i].GetOptions()
+	info, err := proto.GetExtension(opts, cmdproto.E_Info)
+	if err != nil {
+		return "", "", []string{""}, "", cmdproto.FlagDetail_NONE, err
+	}
+	return info.(*cmdproto.FlagDetail).GetName(), info.(*cmdproto.FlagDetail).GetShorthand(), info.(*cmdproto.FlagDetail).GetValue(), info.(*cmdproto.FlagDetail).GetUsage(), info.(*cmdproto.FlagDetail).GetExt(), nil
+}
+
+func FlagsSetup(cmd *cobra.Command, msg interface{}) {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/cmdproto.go]: Would it be possible to restructure this so instead of taking the cmd it returns the cmd, and takes an interface that accepts the `msg` and returns the output + error.

File: pkg/kubectl/cmd/util/cmdproto/cmdproto.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+func extractFlagDetailFromMessage(i int, msg interface{}) (string, string, []string, string, cmdproto.FlagDetail_EXT, error) {
+	_, md := descriptor.ForMessage(msg.(descriptor.Message))
+	opts := md.Field[i].GetOptions()
+	info, err := proto.GetExtension(opts, cmdproto.E_Info)
+	if err != nil {
+		return "", "", []string{""}, "", cmdproto.FlagDetail_NONE, err
+	}
+	return info.(*cmdproto.FlagDetail).GetName(), info.(*cmdproto.FlagDetail).GetShorthand(), info.(*cmdproto.FlagDetail).GetValue(), info.(*cmdproto.FlagDetail).GetUsage(), info.(*cmdproto.FlagDetail).GetExt(), nil
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/cmdproto/cmdproto.go]: nit: split this across multiple lines.

**Comment by pwittrock** [comment]
Review status: 0 of 47 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/cmd/util/cmdproto/cmdproto.go, line 31 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54782#-KyHtWXBnjFNDwtdq7Kl:-KyHtWXBnjFNDwtdq7Km:b-o68d6i) ([raw file](https://github.com/kubernetes/kubernetes/blob/371026467fe6882588eac884ec2000cbf5942a0c/pkg/kubectl/cmd/util/cmdproto/cmdproto.go#L31)):*
> ```Go
> 	"github.com/spf13/cobra"
> 
> 	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
> ```

For v1, lets no create a dependency on any kubernetes/kubernetes packages.  Either drop the functionality, or wrap it in an interface we control in this package and depend upon that.

---

*[pkg/kubectl/cmd/util/cmdproto/cmdproto.go, line 72 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54782#-KyHtvjGNd-yBr6WuK6E:-KyHtvjGNd-yBr6WuK6F:b-qhdk7n) ([raw file](https://github.com/kubernetes/kubernetes/blob/371026467fe6882588eac884ec2000cbf5942a0c/pkg/kubectl/cmd/util/cmdproto/cmdproto.go#L72)):*
> ```Go
> 		Example: templates.Examples(i18n.T(info.GetExample())),
> 		Run: func(cmd *cobra.Command, args []string) {
> 			cmdutil.CheckErr(msg.(CommandExecutor).Complete(f, in, out, err))
> ```

Lets have the CommandExecutor be a separate argument from the message.  That will provide more type safety and flexibility.

---

*[pkg/kubectl/cmd/util/cmdproto/cmdproto.go, line 81 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54782#-KyHuS9A5t-hfOu-B2pF:-KyHuS9A5t-hfOu-B2pG:b8k1rk9) ([raw file](https://github.com/kubernetes/kubernetes/blob/371026467fe6882588eac884ec2000cbf5942a0c/pkg/kubectl/cmd/util/cmdproto/cmdproto.go#L81)):*
> ```Go
> }
> 
> func flagsSetup(cmd *cobra.Command, msg interface{}) {
> ```

lets also add support for parsing some fields from environment variables using a separate set of arguments

---

*[pkg/kubectl/cmd/util/cmdproto/cmdproto.go, line 124 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54782#-KyHuCnBZXAvZovB4D-4:-KyHuCnBZXAvZovB4D-5:b3pn5nf) ([raw file](https://github.com/kubernetes/kubernetes/blob/371026467fe6882588eac884ec2000cbf5942a0c/pkg/kubectl/cmd/util/cmdproto/cmdproto.go#L124)):*
> ```Go
> 			cmd.Flags().StringVarP(&tmpString, d.GetName(), d.GetShorthand(), d.GetValue()[0], d.GetUsage())
> 			v.Field(i).Set(reflect.ValueOf(&tmpString))
> 		case *cmdproto.Array:
> ```

This is great.  Could we also add support for reading a file?  e.g. define a new proto message type for files and allow file flags to be specified.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54782)*

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto (hunk 9 above)
Comment (shiywang) [reply pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: the `enum EXT` is for some mapping variable between cobra and protobuf, for instance if I want to create a cobra `timeDuration` flag, I need to declare an Int64 in proto file and EXT = ISTIME.

**Comment by shiywang** [comment]
@pwittrock still have some unaddressed comments left, will ping you when I'm finished, thanks

**Comment by pwittrock** [comment]
@shiywang Cool.  Thanks.  When you have a chance. lets move this to the kubernetes/kubectl repo.

File: pkg/kubectl/cmd/util/cmdproto/cmdproto.go (hunk 3 above)
Comment (shiywang) [reply pkg/kubectl/cmd/util/cmdproto/cmdproto.go]: Could we do that after kubectl fully segregated from main repo. I think there's a circular dependency problem currently.

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto (hunk 6 above)
Comment (shiywang) [reply pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: done

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto (hunk 7 above)
Comment (shiywang) [reply pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: done

File: pkg/kubectl/cmd/util/cmdproto/cmdproto.go (hunk 11 above)
Comment (shiywang) [reply pkg/kubectl/cmd/util/cmdproto/cmdproto.go]: return a struct instead

File: pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto (hunk 2 above)
Comment (shiywang) [reply pkg/kubectl/cmd/util/cmdproto/k8s_io_kubectl_cmd/cmd.proto]: I can do that for all the scalar value, but default option in protos does not allow you to define a default value for aggregate fields (e.g. no defaults allowed for repeated/map fields or fields whose type is a nested message). wdyt ?

File: pkg/kubectl/cmd/util/cmdproto/cmdproto.go (hunk 3 above)
Comment (pwittrock) [reply pkg/kubectl/cmd/util/cmdproto/cmdproto.go]: We should break those dependencies by defining interfaces and using those.

**Comment by pwittrock** [comment]
Review status: 0 of 27 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/cmd/util/cmdproto/cmdproto.go, line 72 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54782#-KyHtvjGNd-yBr6WuK6E:-KyvP_MHE1sRe7fAsRsA:b-q0rscn) ([raw file](https://github.com/kubernetes/kubernetes/blob/371026467fe6882588eac884ec2000cbf5942a0c/pkg/kubectl/cmd/util/cmdproto/cmdproto.go#L72)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary><blockquote>

Lets have the CommandExecutor be a separate argument from the message.  That will provide more type safety and flexibility.
</blockquote></details>

rather, lets pass in an interface (`type Command interface`) that has these functions so we don't give up type safety.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54782)*
